[elvern18] iP

elvern18 opened the pull request:
# DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. 
It's,
* text-based
* easy to learn
* ~FAST~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/elvern18/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :smile"

And it is **FREE**!

Features:
- [X] Managing tasks
- [ ] Mangaging deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practise Java too. Here's the main method:

```Java
    public static void main(String[] args) throws IOException {
        new Duke(DEFAULT_PATH).run();
    }
```

Commit:
Add Gradle support

elvern18 committed:
completed level 1

elvern18 committed:
completed level 2

elvern18 committed:
Level 3 completed

elvern18 committed:
half completed level 4

elvern18 committed:
completed level 4

elvern18 committed:
partially completed testing

elvern18 committed:
completed A-TextUiTesting

elvern18 committed:
completed Level 5

elvern18 committed:
completed Level-6

elvern18 committed:
specficed import packages instead of wildcare

elvern18 committed:
completed level 8

elvern18 committed:
completed Level-7

elvern18 committed:
changed OS path method

elvern18 committed:
changed output of duke.txt

elvern18 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

elvern18 committed:
merged level 8 branch w master

elvern18 committed:
changed packages imported

elvern18 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

elvern18 committed:
set sdk, change build.gradle application name

Tohtoroo commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+public class Deadline extends Task {
```
Consider adding javadoc?

elvern18 replied:
Done.

Tohtoroo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,178 @@
+
```
Perhaps delete the unneeded white space?

elvern18 replied:
Ok noted!

Tohtoroo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,178 @@
+
+    final static DateTimeFormatter outputFormatter = DateTimeFormatter.ISO_LOCAL_DATE_TIME;
+
+
```
Perhaps delete the unneeded whitespace

Tohtoroo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,178 @@
+        output.close();
     }
 }
+
```
Perhaps delete these unneeded white spaces?

elvern18 replied:
Noted.

Tohtoroo commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+        return String.format("[%s] %s", this.getStatusIcon(), this.description);
+    }
+}
+
```
Perhaps delete the unneeded white space for this line. Only need one white space at the end of each file!

elvern18 replied:
Noted.

Tohtoroo reviewed (commented):
Remember to add javadoc comments! Other than that your code is very readable, and just minor white space problems, but other than these minor stuff, you're good!

elvern18 committed:
completed A-Gradle

weekiat-douze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,176 @@
+                } else if (command.equals(LIST_COMMAND)) {
+                    getList();
+                } else if (isMark(command)) {
+                    handleMark(inputWords);
```
Perhaps the method names here (handle...) can be more descriptive?

elvern18 replied:
Ok, noted!

weekiat-douze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,176 @@
+    public static void handleDelete(String[] inputWords) {
+        int index = Integer.parseInt(inputWords[1]);
+        Task task = list.get(index - 1);
+        list.remove(index-1);
```
You might have missed some while formatting.

weekiat-douze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,176 @@
+    final static DateTimeFormatter outputFormatter = DateTimeFormatter.ISO_LOCAL_DATE_TIME;
+
+
+    public static ArrayList<Task> list;
```
I think the variable name here can be more meaningful.

elvern18 replied:
Ok noted!

weekiat-douze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,176 @@
+        System.out.println(ENTRY);
+        Scanner sc = new Scanner(System.in);
+        list = new ArrayList<Task>();
+        while (true) {
```
Perhaps true can be saved as variable so that it is clear under what conditions this loop will iterate.

elvern18 replied:
Resolved.

weekiat-douze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,176 @@
+        }
+    }
+
+    public static boolean isMark(String word) {
```
I think abstracting these was a great choice, improved readability.

elvern18 replied:
Thank you!

weekiat-douze reviewed (commented):
Looks good. All the best!

elvern18 committed:
refactored and added save option

elvern18 committed:
completed A-MoreOOP

elvern18 committed:
Organised into Packages

elvern18 committed:
reorganised packages

elvern18 committed:
ToDo testing

elvern18 committed:
completed A-Junit

elvern18 committed:
packaged the app as a JAR file

elvern18 committed:
completed JavaDoc for Task

elvern18 committed:
completed JavaDoc for all classes except Duke

elvern18 committed:
madeordering of import statements consistent

elvern18 committed:
find function created

elvern18 committed:
resolved merge conflicts with A-CodingStandard

elvern18 committed:
resolved conflicts with Level-9 branch

elvern18 committed:
Update README.md

elvern18 committed:
Update README.md

elvern18 committed:
Update README.md

Acerizm commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+     * Checks if user wants to unmark a task.
+     *
+     * @param word First word of the user command.
+     * @return Yes or no in the form of a boolean, where true is yes and false is no.
+     */
+    public boolean isUnmark(String word) {
+        return word.equals(UNMARK_COMMAND);
+    }
+
+    /**
+     * Checks if it is a ToDo command.
+     *
+     * @param word First word of the user command.
+     * @return Yes or no in the form of a boolean, where true is yes and false is no.
+     */
+    public boolean isToDo(String word) {
+        return word.equals(TODO_COMMAND);
+    }
+
+    /**
+     * Checks if it is a Deadline command.
+     *
+     * @param word First word of the user command.
+     * @return Yes or no in the form of a boolean, where true is yes and false is no.
+     */
+    public boolean isDeadline(String word) {
+        return word.equals(DEADLINE_COMMAND);
+    }
+
+    /**
+     * Checks if it is a Event command.
+     *
+     * @param word First word of the user command.
+     * @return Yes or no in the form of a boolean, where true is yes and false is no.
+     */
+    public boolean isEvent(String word) {
+        return word.equals(EVENT_COMMAND);
+    }
+
+}
```
Looks good and clean! Keep it up!

Acerizm commented on src/main/java/tasklist/TaskList.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Encapsulates the list of tasks inputted by the user.
+ */
+public class TaskList {
+    public ArrayList<Task> list;
```
I would suggest maybe making the ArrayList<Task> list modifier to be private so that only TaskList can access the list! The rest looks good!

Acerizm commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+     * @param ui User Interface .
+     * @param list List of tasks.
+     */
+    public void parse(String input, Ui ui, TaskList list) {
+        try {
+            String[] inputWords = input.split(" ", 2);
+            String command = inputWords[0];
+            if (command.equals(EXIT_COMMAND)) {
+                ui.printBye();
+                ui.isClosed = true;
+            } else if (command.equals(LIST_COMMAND)) {
+                ui.printGetList(list);
+            } else if (command.equals(FIND_COMMAND)) {
+                this.handleFind(inputWords, ui, list);
+            } else if (isMark(command)) {
+                this.handleMark(inputWords, ui, list);
+            } else if (isUnmark(command)) {
+                this.handleUnmark(inputWords, ui, list);
+            } else if (isToDo(command)) {
+                this.handleToDo(inputWords, ui, list);
+            } else if (isDeadline(command)) {
+                this.handleDeadline(inputWords, ui, list);
+            } else if (isEvent(command)) {
+                this.handleEvent(inputWords, ui, list);
+            } else if (command.equals(DELETE_COMMAND)) {
+                this.handleDelete(inputWords, ui, list);
+            } else {
+                throw new InvalidInputException();
+            }
+        } catch (Exception e) {
+            ui.printException(e);
+        }
+    }
```
Good code quality for avoiding the arrowhead style code! Keep it up!

Acerizm reviewed (commented):
Code is very clean and many features have been extracted out perfectly! Keep up the good work :)

elvern18 committed:
completed till part 3

elvern18 committed:
completed part 4 of lvl 10

elvern18 committed:
completed checkstyle

elvern18 committed:
merged with level 10

elvern18 committed:
removed unnecessary packages

elvern18 committed:
trying to make branch-level-8 detectable

elvern18 committed:
trying to make level 10 branch detectable

elvern18 committed:
changed mainClassName in build.gradle to Launcher

elvern18 committed:
minor changes

elvern18 committed:
refactored GUI to respond to user's input

elvern18 committed:
improved coding quality to follow coding standards

elvern18 committed:
changed dukeexceptions package name

elvern18 committed:
Implement assertions to Storage and TaskList files

Developers will not know the correctness of the assumption that the task
is not null, or the size of the list of tasks is negative.

Neglecting these assumptions can cause severe errors and even crash of
chatbot.

Let's use assertions to ensure that these assumptions are checked.

This prevents severe errors from occuring to the users, and even
crashing of chatbot.

Assertions are used to ensure our test conditions are true.

elvern18 committed:
Merge pull request #3 from elvern18/branch-A-Assertions

Implement assertions to Storage and TaskList files

elvern18 committed:
Merge branch 'master' into branch-A-CodeQuality

elvern18 committed:
Merge pull request #2 from elvern18/branch-A-CodeQuality

Made changes to adhere to Coding Standards

elvern18 committed:
Friendlier Syntax: Make input required more user friendly

Commands have to be spelt in full.

A misspelling of the first word of the whole input will cause an error
to be received.

A shorter input will make it more flexible for the user to communication
with the chatbot on the instruction that they want the chatbot to
execute.

Users now can use a single letter to execute the type of command to the
chatbot. This reduces room for misspelling.

elvern18 committed:
made some minor changes as well

elvern18 committed:
made javadocs changes

elvern18 committed:
Update README.md

elvern18 committed:
Update README.md

elvern18 committed:
Update README.md

elvern18 committed:
Update README.md

elvern18 committed:
Update README.md

elvern18 committed:
final changes to UI and Exception output

elvern18 committed:
Merge branch 'master' of https://github.com/elvern18/ip

elvern18 committed:
Added Ui.png via upload

elvern18 committed:
added additional exception for datetime format for deadline task

elvern18 committed:
Merge branch 'master' of https://github.com/elvern18/ip

elvern18 committed:
Update README.md

elvern18 committed:
Update README.md

elvern18 committed:
minor changes to code quality

elvern18 committed:
edited storage to make directory / file if not present

elvern18 committed:
minor changes to coding standards
